### PR TITLE
test(client): overhaul e2e testing story to use cypress

### PR DESCRIPTION
**Pull request by PascalPrecht**
TODOS:

- [x] - Figure out environment story (e.g. everyone should be able to run e2e against local server)

### CODE REVIEW

**Comment by d3lm** [comment]
Is this still a WIP or ready for review?

**Comment by PascalPrecht** [comment]
I personally treat any WIP PR as ready for review.

In this case we could certainly provide more e2e tests and CI integration is also still in flux. But please feel free to leave feedback already.

**Comment by PascalPrecht** [comment]
@cburgdorf In our last discussion about this I think we've decided to let the e2e tests simply run against our staging system. That's fine by me. However we'll have cases where we run into chicken and egg problems. For example, to have the current e2e test suite pass, I needed to make some changes in our app's template.

Those changes are part of this PR but not yet deployed to staging, making our e2e tests naturally fail. I guess we regularly need to keep an eye on this, when a PR requires a deployment to staging first.

That being said, would you mind putting this branch on staging so I can work on the CI integration?

**Comment by PascalPrecht** [comment]
Here's another thing we need to agree on: Environment variables.

Basically what we want is:

- Everyone should be able to run e2e tests locally on a local server (most likely `localhost:4200`)
- e2e tests should be executed against staging.machinelabs.ai on CI
- Maybe, in the future we want to be able to run e2e tests against different environments

Cypress offers a few ways to deal with this:

**Using cypress.json**

This would be a general environment configuration that would also be part of version control. Ideally, here we would configure default settings that should be the same for everyone. E.g. `baseUrl = localhost:4200`, no screenshots, no video recording.

**Creating a `cypress.env.json`**

This file would not be part of version control and everyone would basically be able to create their own custom environment settings. Also, conflicting values between `cypress.env.json` and `cypress.json` will be overridden. You can see the env file as a preferred setting with `cypress.json` being the fallback.

**Using `CYPRESS_*` environment variables**

Basically environment variables that can be quickly set/overridden. Those will also override values in `cypress.env.json` and `cypress.json`. Good for CI environments.

**Using CLI's `--env` option**

Similar to Angular CLI's `env` option, there's an `--env` flag that takes a comma separated list of values for environment variables. Those have the highest priority over all other options.

More on all those options here: https://docs.cypress.io/guides/guides/environment-variables.html

### Proposal

In order let everyone run with his/her own config, but at the same time sharing the same for e.g. CI, I'd suggest we do the following:

- We keep the already introduced `cypress.json` which sets `baseUrl: localhost:4200` and turns off screenshots and video

- Anyone who'd like to override any default settings can introduce their own `cypress.env.json` file (which we'll ignore in the project), or use any of the other options respectively.

- Anything else that needs change for CI can be set in environment variables. However, right now, I think the provided defaults in `cypress.json` would do the trick.

**Comment by cburgdorf** [comment]
Quick summary of our video call:

1. The e2e tests should run against a local webserver serving the client as it is in this PR. In other words, the e2e tests should *not* run against http://staging.machinelabs.ai because otherwise any PR that contains a new e2e test will fail in CI by definition as it just doesn't exist on the staging system yet.

2. However, the client that will run on a local webserver will run against the rest of the staging system (server, firebase). That's a bit of a compromise so that we don't have to have a system where each PR deploys it's own atomic universe (something that isn't trivial to achieve) at the cost of being dependent against the version of the server and the firebase that is currently deployed on staging. 

3. @PascalPrecht mentined that running cypress on top of a `ng serve` is problematic so we try out a different approach where we first generated a dist and serve that manually without using the angular-cli.

4. Regarding the environment variables. I'm not sure yet how many environment variables we need. I'm fine to go with any of the proposed options.

**Comment by PascalPrecht** [comment]
Thanks for the write-up @cburgdorf !

> @PascalPrecht mentined that running cypress on top of a ng serve is problematic so we try out a different approach where we first generated a dist and serve that manually without using the angular-cli.

Here's the related issue: https://github.com/bahmutov/start-server-and-test/issues/4#issuecomment-376971830

**Comment by PascalPrecht** [comment]
Alright, updated the PR to now run cypress on CI as well. Let's see what travis says.

File: client/cypress/integration/editor-view_spec.ts [hunk 1]
```diff
@@ -0,0 +1,172 @@
...
+
+      it('should allow users to edit file names', () => {
+        const fileIndex = 2;
+        const newName = 'changed.py';
+
+        editorView.toggleFileTree();
+        editorView.addFile('ml-main-2.py')
+        cy.wait(1000);
+        editorView.changeFileName(fileIndex, newName);
+        cy.wait(1000);
```
Comment (PascalPrecht) [inline client/cypress/integration/editor-view_spec.ts]: Those `wait()` calls are more a workaround for now until we figure out how to handle this in a nice way...

File: client/cypress/support/po.ts [hunk 2]
```diff
@@ -0,0 +1,177 @@
+export class EditorViewPageObject {
```
Comment (PascalPrecht) [inline client/cypress/support/po.ts]: I guess, once we introduce more PageObjects, we can move those into dedicated files.

File: client/package.json [hunk 3]
```diff
@@ -4,19 +4,23 @@
   "license": "MIT",
   "angular-cli": {},
   "scripts": {
-    "update-shared": "(cd ../shared && node build.js)",
     "start": "ng serve",
+    "build": "ng build --prod",
+    "build:staging": "ng build --env=staging --prod --sourcemaps false --progress false",
```
Comment (PascalPrecht) [inline client/package.json]: Turning off `sourcemaps` here as we don't need those in production builds. Speeds up the build process.

File: client/package.json [hunk 4]
```diff
@@ -4,19 +4,23 @@
...
     "format:base": "npm run prettier -- '{src,e2e}/**/*{.ts,.js,.json,.css,.scss}'",
     "format:check": "npm run format:base -- --list-different",
     "format:fix": "npm run format:base -- --write",
-    "ci": "npm run lint && npm run format:check && npm run test && npm run test:e2e:production",
-    "test": "ng test --single-run && ng e2e --env=e2e",
-    "test:e2e:production": "ng e2e --target=production --env=e2e",
-    "pree2e": "webdriver-manager update",
-    "e2e": "ng e2e --serve false",
-    "node_modules": "rm -rf ./node_modules && yarn install"
+    "ci": "npm run lint-and-check && npm run test:single-run && CYPRESS_baseUrl=http://localhost:5000 start-server-and-test build-and-serve:staging http://localhost:5000 e2e",
```
Comment (PascalPrecht) [inline client/package.json]: Notice that I'm overriding cypress `baseUrl` here with an environment variable. This is to ensure that in the `ci` script, cypress looks for `localhost:5000`.

There might be a way to configure env variables in travis, but then the `ci` script wouldn't work locally anymore, unless the env variable is set manually.

No strong opinion here.

**Pull request merged**